Update ansible-logstash-callback

ujenmr opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
plugins/callback/logstash.py

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
LOGSTASH_SERVER=127.0.0.1 LOGSTASH_PORT=5000 ansible-playbook playbook.yml
```

```paste below
{
              "ansible_type" => "start",
                     "level" => "INFO",
                   "session" => "111f73e0-eb57-11e6-bc8d-e4115b24f077",
                   "message" => "START playbooks/deploy-aws-bastion.yml",
                      "type" => "ansible",
          "ansible_playbook" => "playbooks/deploy-aws-bastion.yml",
                      "tags" => [],
                      "path" => "/opt/work/ansible-playbook-bastion/ansible/plugins/callbacks/logstash-latest.py",
                "@timestamp" => 2017-02-05T03:56:26.335Z,
                      "port" => 43920,
                  "@version" => "1",
                      "host" => "laptappy",
               "logger_name" => "python-logstash-logger",
"ansible_pre_command_output" => "ansible 2.2.1.0",
                    "status" => "OK"
}
{
              "ansible_type" => "task",
                     "level" => "INFO",
            "ansible_result" => "{\"changed\": false, \"cmd\": \"TZ=':US/Eastern' date +%s\", \"delta\": \"0:00:00.003644\", \"end\": \"2017-02-05 03:56:57.823829\", \"rc\": 0, \"start\": \"2017-02-05 03:56:57.820185\", \"stderr\": \"\", \"stdout\": \"1486267017\", \"stdout_lines\": [\"1486267017\"], \"warnings\": []}",
                   "session" => "111f73e0-eb57-11e6-bc8d-e4115b24f077",
           "ansible_changed" => false,
                   "message" => "{\"changed\": false, \"cmd\": \"TZ=':US/Eastern' date +%s\", \"delta\": \"0:00:00.003644\", \"end\": \"2017-02-05 03:56:57.823829\", \"rc\": 0, \"start\": \"2017-02-05 03:56:57.820185\", \"stderr\": \"\", \"stdout\": \"1486267017\", \"stdout_lines\": [\"1486267017\"], \"warnings\": []}",
                      "type" => "ansible",
          "ansible_playbook" => "playbooks/deploy-aws-bastion.yml",
              "ansible_task" => "Get post-epoch time",
                      "tags" => [],
                      "path" => "/opt/work/ansible-playbook-bastion/ansible/plugins/callbacks/logstash-latest.py",
              "ansible_host" => "172.17.0.27",
                "@timestamp" => 2017-02-05T03:56:57.855Z,
                      "port" => 43920,
                      "host" => "laptappy",
                  "@version" => "1",
               "logger_name" => "python-logstash-logger",
"ansible_pre_command_output" => "ansible 2.2.1.0",
                    "status" => "OK"
}
{
              "ansible_type" => "finish",
                     "level" => "INFO",
            "ansible_result" => "{\"172.17.0.27\": {\"unreachable\": 0, \"skipped\": 2, \"ok\": 15, \"changed\": 0, \"failures\": 0}}",
                   "session" => "111f73e0-eb57-11e6-bc8d-e4115b24f077",
                   "message" => "{\"172.17.0.27\": {\"unreachable\": 0, \"skipped\": 2, \"ok\": 15, \"changed\": 0, \"failures\": 0}}",
                      "type" => "ansible",
          "ansible_playbook" => "playbooks/deploy-aws-bastion.yml",
                      "tags" => [],
                      "path" => "/opt/work/ansible-playbook-bastion/ansible/plugins/callbacks/logstash-latest.py",
                "@timestamp" => 2017-02-05T03:56:57.857Z,
                      "port" => 43920,
                  "@version" => "1",
                      "host" => "laptappy",
               "logger_name" => "python-logstash-logger",
"ansible_pre_command_output" => "ansible 2.2.1.0",
                    "status" => "OK"
}```

Andersson007 commented:
@ujenmr i don't use this plugin. Don't the changes break backward compatibility anyhow?

felixfontein commented:
This looks like a huge refactoring. It also (re-)introduce a lot of things that shouldn't be there. Is there a reason you are doing such a massive change? Also, this change definitely breaks backwards compatibility, since the log messages change.

Andersson007 commented:
Needs to have felixfontein's questions answered 
needs_info

ujenmr commented:
> This looks like a huge refactoring. It also (re-)introduce a lot of things that shouldn't be there. Is there a reason you are doing such a massive change? Also, this change definitely breaks backwards compatibility, since the log messages change.

These changes are just optimization/refactoring, I tried to maintain compatibility with older versions

Andersson007 commented:
Can this changes possibly change the current behavior of existing options, i.e. users will get something different from what they have now?

ujenmr commented:
> Can this changes possibly change the current behavior of existing options, i.e. users will get something different from what they have now?

no, for example, the main field is ansible_result, just look on, behavior not changed, field contains self._dump_results(result._result) same as in the previous version

Andersson007 commented:
@ujenmr ok, thanks for the explanation, i'll review this quickly

Andersson007 commented:
```
logstash callback - migrate to python3-logstash
```
1) is this a requirement?
2) is this compatible with python 2?

ujenmr commented:
I just checked, and all is ok on python 2

```
ansible-playbook 2.8.14
  config file = /Users/ujen/Documents/workspace/ansible/platform/ansible.cfg
  configured module search path = [u'/Users/ujen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/ujen/.pyenv/versions/2.7.18/lib/python2.7/site-packages/ansible
  executable location = /Users/ujen/.pyenv/versions/2.7.18/bin/ansible-playbook
  python version = 2.7.18 (default, Oct 20 2020, 11:50:25) [GCC Clang 11.0.0]
```

Andersson007 commented:
> I just checked, and all is ok on python 2
> 
> ```
> ansible-playbook 2.8.14
>   config file = /Users/ujen/Documents/workspace/ansible/platform/ansible.cfg
>   configured module search path = [u'/Users/ujen/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
>   ansible python module location = /Users/ujen/.pyenv/versions/2.7.18/lib/python2.7/site-packages/ansible
>   executable location = /Users/ujen/.pyenv/versions/2.7.18/bin/ansible-playbook
>   python version = 2.7.18 (default, Oct 20 2020, 11:50:25) [GCC Clang 11.0.0]
> ```

Ok, thanks for the info! I left a couple of comments

@felixfontein what's your opinion? @ujenmr is an author of the module and the PR doesn't break anything and is compatible with Python 2 as he said.

felixfontein commented:
@Andersson007 I'm fine with merging it once it is cleaned up (especially with regard to all the modified/removed/added version_added statements).

Would be interesting to hear @chris93111's opinion on this though, since they worked on the module recently (#610).

Andersson007 commented on plugins/callback/logstash.py:
```diff
@@ -1,57 +1,86 @@
+DOCUMENTATION = r'''
     callback: logstash
     type: notification
+    author: Yevhen Khmelenko <ujenmr@gmail.com>
```
```suggestion
    author: Yevhen Khmelenko (@ujenmr)
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -1,57 +1,86 @@
     short_description: Sends events to Logstash
     description:
       - This callback will report facts and task events to Logstash https://www.elastic.co/products/logstash
+    version_added: "2.3"
```
```suggestion
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -1,57 +1,86 @@
         ini:
           - section: callback_logstash
             key: server
-            version_added: 1.0.0
```
This version_added needs to stay.

ujenmr replied:
added with https://github.com/ansible-collections/community.general/pull/641/commits/f32be815f0cb7f37bec0e72d752339d7fa709792

felixfontein replied:
The version_added is for the `ini` key, not for the option.

ujenmr replied:
like this https://github.com/ansible-collections/community.general/pull/641/commits/4d5a0e7c341162b378f60b1aec820541c42ff40e ?

felixfontein replied:
Yes, but now it's gone again. Please make sure that the diff for this PR doesn't remove anything for options you are not adding. Reordering keys is ok (though why have `env` after `ini`?), but not removing things.

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -1,57 +1,86 @@
         ini:
           - section: callback_logstash
             key: port
-            version_added: 1.0.0
```
Same here.

ujenmr replied:
added with https://github.com/ansible-collections/community.general/pull/641/commits/f32be815f0cb7f37bec0e72d752339d7fa709792

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -1,57 +1,86 @@
+            key: pre_command
+        env:
+          - name: LOGSTASH_PRE_COMMAND
+        version_added: "2.6"
```
```suggestion
        version_added: "1.3.0"
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -1,57 +1,86 @@
         ini:
           - section: callback_logstash
             key: type
-            version_added: 1.0.0
```
This also needs to stay.

ujenmr replied:
added with https://github.com/ansible-collections/community.general/pull/641/commits/f32be815f0cb7f37bec0e72d752339d7fa709792

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -1,57 +1,86 @@
+ansible.cfg: >
+    # Enable Callback plugin
+    [defaults]
+        callback_whitelist = logstash
```
```suggestion
        callback_whitelist = community.general.logstash
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -63,76 +92,85 @@
     CALLBACK_VERSION = 2.0
     CALLBACK_TYPE = 'aggregate'
-    CALLBACK_NAME = 'community.general.logstash'
+    CALLBACK_NAME = 'logstash'
```
```suggestion
    CALLBACK_NAME = 'community.general.logstash'
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -63,76 +92,85 @@
-        self.hostname = socket.gethostname()
-        self.session = str(uuid.uuid1())
-        self.errors = 0
+        try:
+            self.ls_server = self.get_option('server')
+        except KeyError:
+            self.ls_server = "localhost"
```
```suggestion
        self.ls_server = self.get_option('server')
```
Same for the others below.

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -63,76 +92,85 @@
+        data['ansible_type'] = "start"
+        data['status'] = "OK"
+        data['ansible_playbook'] = playbook._file_name
+        self.logger.info("START PLAYBOOK | %s", data['ansible_playbook'], extra=data)
```
You are changing the text that is logged. Please remove that from this PR, and add it to another PR (that will be a breaking change).

(Same for the similar lines.)

ujenmr replied:
I plan to add property `format_version` for switching new format (like "START PLAYBOOK | bla bla") and legacy

ujenmr replied:
format can be v1 or v2 (default will be v1)

felixfontein replied:
That works for me!

chris93111 commented:
Hi I can test in a production environment and get back to you if you want ?

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,79 @@
         ini:
           - section: callback_logstash
             key: server
-            version_added: 1.0.0
```
This line needs to be re-added.

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,79 @@
+        env:
+          - name: LOGSTASH_PORT
         default: 5000
+        version_added: 1.0.0
```
This needs to be removed, and the `version_added: 1.0.0` above needs to be re-added. (This is at the wrong level.)

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,79 @@
+            key: pre_command
+        env:
+          - name: LOGSTASH_PRE_COMMAND
+        version_added: "1.0.0"
```
If this is a new option, it needs to be
```suggestion
        version_added: "1.3.0"
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,79 @@
+        env:
+          - name: LOGSTASH_TYPE
         default: ansible
+        version_added: 1.0.0
```
This also needs to be at another level.

Please look at the diff and restore the original state: https://github.com/ansible-collections/community.general/pull/641/files

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,79 @@
+        ini:
+          - section: callback_logstash
+            key: format_version
+        default: v1
```
I assume you want:
```suggestion
        default: v1
        type: str
        choices:
          - v1
          - v2
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,79 @@
+        default: v1
+        env:
+          - name: LOGSTASH_FORMAT_VERSION
+        version_added: 1.0.0
```
```suggestion
        version_added: 1.3.0
```

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -63,76 +102,96 @@
-            self._display.warning("The required python-logstash is not installed. "
-                                  "pip install python-logstash")
+            self._display.warning("The required python3-logstash is not installed. "
+                                  "pip install python3-logstash")
```
I guess you should mention that `python-logstash` should be used for Python 2, and `python3-logstash` for Python 3?

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -63,76 +102,96 @@
-        self.hostname = socket.gethostname()
-        self.session = str(uuid.uuid1())
-        self.errors = 0
+        try:
```
The `try`/`except` is not needed. Please remove it. (Same below.)

ujenmr commented:
> Hi i can test in a production environment and get back to you if you want ? (python2/3)

Sure it will be cool

Andersson007 commented:
Cool, let's wait @chris93111 's feedback. Thank you!

Andersson007 commented:
Also CI tests have failed
```
 docker_stack_task_info-_TMzqD / /root/ansible_collections/community/general/tests/output/.tmp/integration/docker_stack_task_info-W7Yu45-ÅÑŚÌβŁÈ/tests/integration/targets/docker_stack_task_info/tasks/test_stack_task_info.yml:72 / [testhost] testhost: docker_stack_task_info : assert stack facts that=[u'output.results | type_debug == "list"', u'output.results[0].DesiredState == "Running"', u'output.results[0].Image == "busybox:latest"', u'output.results[0].Name == "test_stack_busybox.1"']
 failure: Assertion failed

{
    "assertion": "output.results[0].DesiredState == \"Running\"", 
    "changed": false, 
    "evaluated_to": false, 
    "msg": "Assertion failed"
}
```

felixfontein commented:
@Andersson007 that CI error is unrelated. The `docker_stack_task_info` tests unfortunately sometimes fail.

chris93111 commented:
Hi i test in my environment, its not working with "self.handler = logstash.LogstashHandler" with python2 or 3, its working with TCPLogstashHandler

python-logstash == 0.4.6

Andersson007 commented:
@ujenmr ^

ujenmr commented:
@chris93111 can you try test again

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,85 @@
         description: Address of the Logstash server
-        env:
-          - name: LOGSTASH_SERVER
+        version_added: 1.0.0
```
Please fix the existing `version_added` entries. Your PR will never get merged if you don't do it. Seriously!

ujenmr replied:
I removed `version added` entries from old sections, please check one more time

felixfontein replied:
Please take a look at the diff yourself: https://github.com/ansible-collections/community.general/pull/641/files

As long as you can see that version_added has been removed or moved around, you need to fix that. I really don't want to invest any more time in this.

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -18,40 +18,85 @@
         default: 5000
+      pre_command:
+        description: Executes command before run and result put to ansible_pre_command_output field.
+        version_added: "1.3.0"
```
Next release will be 2.0.0, so you need to update the `version_added` for the new options:
```suggestion
        version_added: "2.0.0"
```

chris93111 commented:
Ok i try and send feedback ;)

chris93111 commented:
work ;)

Andersson007 reviewed (commented):
@ujenmr thanks for your effort. Please look at @felixfontein 's requests about `version_added`

ujenmr commented:
@Andersson007 thank you for review, I fixed all requests about `version_added` and rebased changes, @felixfontein can you check review one more time, I hope all will be good

felixfontein commented:
@ujenmr I don't know how the diff view looks for you, but for me it is still obvious that it is wrong, since you modify options this PR should not be touching and move `version_added` in it around:
![image](https://user-images.githubusercontent.com/5781356/104181154-928e3280-540e-11eb-83cb-0e4a728525d9.png)

Same for the `port` and `type` options.

ujenmr commented:
@felixfontein did, please check I hope this is the last time

felixfontein commented:
@ujenmr I still see some `version_added` values that were moved to another location in the YAML docs: 
![image](https://user-images.githubusercontent.com/5781356/105245074-ae61a900-5b71-11eb-8861-7e0a510d2719.png)

ujenmr commented:
@felixfontein, I fixed wrong tag location, if you think that it is not necessary to do this, then I will update soon

ujenmr commented:
@felixfontein, done, please check

felixfontein commented:
> @felixfontein, I fixed wrong tag location, if you think that it is not necessary to do this, then I will update soon

The problem is that it wasn't wrong, but correct: the INI configuration key has been added in version 1.0.0, not the whole option. That was already there before. You moved the `version_added` to the option, thus declaring the option as "added in community.general 1.0.0", which is wrong.

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -43,15 +43,60 @@
+'''
+
+EXAMPLES = r'''
+ansible.cfg: >
```
```suggestion
ansible.cfg: |
```
Otherwise you're telling a YAML parser to remove all single newlines :)

felixfontein commented on plugins/callback/logstash.py:
```diff
@@ -43,15 +43,60 @@
+        pre_command = git rev-parse HEAD
+        type = ansible
+
+11-input-tcp.conf: >
```
```suggestion
11-input-tcp.conf: |
```

felixfontein commented on changelogs/fragments/641-update-ansible-logstash-callback.yml:
```diff
@@ -0,0 +1,3 @@
+minor_changes:
+- logstash callback - migrate to python3-logstash (https://github.com/ansible-collections/community.general/pull/641).
+- logstash callback - improve logstash message structure (https://github.com/ansible-collections/community.general/pull/641).
```
```suggestion
- logstash callback - improve logstash message structure, needs to be enabled with the ``format_version`` option (https://github.com/ansible-collections/community.general/pull/641).
```

felixfontein commented:
@Andersson007 @chris93111 any more comments? 2.0.0 will be released next week, if nobody complains I'll merge by then.

Andersson007 reviewed (approved):
I don't have any comments

chris93111 commented:
ok for me

felixfontein commented:
@ujenmr thanks for implementing this!
@chris93111 @Andersson007 thanks for testing and reviewing!
